crates/dbsp/src/operator/neighborhood.rs

```diff
@@ -0,0 +1,696 @@
+                reset,
+                |range, delta, reset: &bool| {
+                    if *reset {
+                        range
+                    } else {
+                        delta.clone()
+                    }
+                },
+            )
+        })
+    }
+}
+
+/// Computes the neighborhood without row numbers.
+///
+/// Used for per-worker neighborhood computation in each worker, where
+/// row numbers are not needed (since row numbers are assigned by
+/// worker 0).
+///
+/// The operator takes a trace of the input table and produces
+/// a complete neighborhood as it output.  The internal implementation
+/// is non-incremental and will do the work proportional to the size
+/// of the neighborhood even if the neighborhood has not
+/// changed since the last clock cycle.  This is ok assuming small
+/// neighborhoods.  We may want to switch to an incremental
+/// implementation in the future.
+struct NeighborhoodLocal<T>
+where
+    T: BatchReader,
+{
+    _phantom: PhantomData<T>,
+}
+
+impl<T> NeighborhoodLocal<T>
+where
+    T: BatchReader,
+{
+    fn new() -> Self {
+        Self {
+            _phantom: PhantomData,
+        }
+    }
+}
+
+impl<T> Operator for NeighborhoodLocal<T>
+where
+    T: BatchReader + 'static,
+{
+    fn name(&self) -> Cow<'static, str> {
+        Cow::from("NeighborhoodLocal")
+    }
+    fn fixedpoint(&self, _scope: Scope) -> bool {
+        true
+    }
+}
+
+impl<T>
+    BinaryOperator<
+        T,
+        Option<NeighborhoodDescr<T::Key, T::Val>>,
+        OrdIndexedZSet<T::Key, T::Val, T::R>,
+    > for NeighborhoodLocal<T>
+where
+    T: BatchReader<Time = ()>,
+    T::Key: DBData,
+    T::Val: DBData,
+    T::R: DBWeight,
+{
+    fn eval<'a>(
+        &mut self,
+        input_trace: &T,
+        descr: &Option<NeighborhoodDescr<T::Key, T::Val>>,
+    ) -> OrdIndexedZSet<T::Key, T::Val, T::R> {
+        let mut cursor = input_trace.cursor();
+
+        if let Some(descr) = descr {
+            let (anchor_key, anchor_val) = &descr.anchor;
+
+            // Forward pass: locate the anchor and `decr.after`
+            // following rows.
+            let mut after = Vec::with_capacity(descr.after);
+            let mut offset = 0;
+
+            cursor.seek_keyval(anchor_key, anchor_val);
+            while cursor.keyval_valid() && offset < descr.after {
+                let w = cursor.weight();
+
+                if !cursor.weight().is_zero() {
+                    after.push(((cursor.key().clone(), cursor.val().clone()), w));
+                    offset += 1;
+                }
+                cursor.step_keyval();
+            }
+
+            // Reverse pass: find `descr.before` rows preceding the anchor.
+            cursor.fast_forward_keys();
+            cursor.fast_forward_vals();
+
+            let mut before = Vec::with_capacity(descr.before);
+            offset = 1;
+
+            cursor.seek_keyval_reverse(anchor_key, anchor_val);
```

Comment:
couldn't you cache the previous position somehow?

Reply:
We don't have the machinery to do this at the moment (and it's not completely trivial for, e.g., list cursors), but it's not a big deal in this case since we're talking about one extra lookup per operator invocation.